Allow Union of scalars to be NumberType

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68773
* #68772
* #66592
* __->__ #66591

Differential Revision: [D31632599](https://our.internmc.facebook.com/intern/diff/D31632599)

tugsbayasgalan committed:
Aloow Union of scalars to be NumberType

[ghstack-poisoned]

gmagogsfm reviewed (commented):
Could you add a test? Thanks

tugsbayasgalan committed:
Update on "Allow Union of scalars to be NumberType"


Differential Revision: [D31632599](https://our.internmc.facebook.com/intern/diff/D31632599)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31632599).

tugsbayasgalan commented on test/test_jit.py:
```diff
@@ -11125,6 +11125,12 @@ def fn(x, y):
+        @torch.jit.script
+        def fn(x: Union[int, complex, float], y: Union[int, complex, float]):
+            return x + y
+        FileCheck().check(": Scalar):").run(fn.graph)
```
https://github.com/pytorch/pytorch/blob/e2be087207c210a30e201b786fc4d3aab8336837/torch/csrc/jit/frontend/schema_type_parser.cpp#L57 (This is how its represented in the frontend I think)

gmagogsfm commented on torch/jit/annotations.py:
```diff
@@ -411,6 +414,7 @@ def ann_to_type(ann, loc):
     'TensorType',
     'TupleType',
     'FloatType',
+    'NumberType',
```
If it is only about a possible future use case, then let's not add it now. 

Fun read about YAGNI:
https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

tugsbayasgalan committed:
Update on "Allow Union of scalars to be NumberType"


Differential Revision: [D31632599](https://our.internmc.facebook.com/intern/diff/D31632599)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31632599).

tugsbayasgalan committed:
Update on "Allow Union of scalars to be NumberType"


Differential Revision: [D31632599](https://our.internmc.facebook.com/intern/diff/D31632599)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31632599).

tugsbayasgalan committed:
Update on "Allow Union of scalars to be NumberType"


Differential Revision: [D31632599](https://our.internmc.facebook.com/intern/diff/D31632599)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Allow Union of scalars to be NumberType"


Differential Revision: [D31632599](https://our.internmc.facebook.com/intern/diff/D31632599)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Allow Union of scalars to be NumberType"


Differential Revision: [D31632599](https://our.internmc.facebook.com/intern/diff/D31632599)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31632599).
